test/extensions/stats_sinks/metrics_service/metrics_service_integration_test.cc

```diff
@@ -56,45 +56,45 @@ class MetricsServiceIntegrationTest : public HttpIntegrationTest,
   }
 
   void waitForMetricsRequest() {
-    envoy::service::metrics::v2::StreamMetricsMessage request_msg;
-    metrics_service_request_->waitForGrpcMessage(*dispatcher_, request_msg);
-    EXPECT_STREQ("POST", metrics_service_request_->headers().Method()->value().c_str());
-    EXPECT_STREQ("/envoy.service.metrics.v2.MetricsService/StreamMetrics",
-                 metrics_service_request_->headers().Path()->value().c_str());
-    EXPECT_STREQ("application/grpc",
-                 metrics_service_request_->headers().ContentType()->value().c_str());
-    EXPECT_TRUE(request_msg.envoy_metrics_size() > 0);
-    const Protobuf::RepeatedPtrField<::io::prometheus::client::MetricFamily>& envoy_metrics =
-        request_msg.envoy_metrics();
+    bool known_histogram_exists = false;
     bool known_counter_exists = false;
     bool known_gauge_exists = false;
-    bool known_histogram_exists = false;
-    for (::io::prometheus::client::MetricFamily metrics_family : envoy_metrics) {
-      if (metrics_family.name() == "cluster.cluster_0.membership_change" &&
-          metrics_family.type() == ::io::prometheus::client::MetricType::COUNTER) {
-        known_counter_exists = true;
-        EXPECT_EQ(1, metrics_family.metric(0).counter().value());
-      }
-      if (metrics_family.name() == "cluster.cluster_0.membership_total" &&
-          metrics_family.type() == ::io::prometheus::client::MetricType::GAUGE) {
-        known_gauge_exists = true;
-        EXPECT_EQ(1, metrics_family.metric(0).gauge().value());
-      }
-      if (metrics_family.name() == "cluster.cluster_0.upstream_rq_time" &&
-          metrics_family.type() == ::io::prometheus::client::MetricType::SUMMARY) {
-        known_histogram_exists = true;
-        Stats::HistogramStatisticsImpl empty_statistics;
-        EXPECT_EQ(metrics_family.metric(0).summary().quantile_size(),
-                  empty_statistics.supportedQuantiles().size());
-      }
-      ASSERT(metrics_family.metric(0).has_timestamp_ms());
-      if (known_counter_exists && known_gauge_exists && known_histogram_exists) {
-        break;
+    while (known_counter_exists && known_gauge_exists && known_histogram_exists) {
```

Comment:
maybe bound this to a fixed number of loops or time-bound so unit-tests don't hang or spin if there's a problem.

Reply:
I think it will timeout if not found...

Reply:
I just tried running 'bazel test' where I injected a "while (true) {}" in a test method, and it seems to be running for a very long time.   What do you expect would cause such a timeout?

Would you lose anything by (say) having this loop terminate after 100 iterations?

Reply:
My infinite-loop test finally timed out after 5 minutes.  I bet you can make the test tighter than that.  However I'd be happy too if we actually plan to fix this robustly in the future, with a TODO.

Reply:
I think that makes sense. I have added TODO to fix it in a more robust way

Reply:
We don't have to fix it here, but I think just setting a reasonable-but-not-ridiculous timeout, like a 20 second timer, would be the right way to make this more robust.